tiledb/sm/query/reader.cc

```diff
@@ -353,17 +353,25 @@ Status Reader::read() {
       zero_out_buffer_sizes();
       RETURN_NOT_OK(read_state_.split_current());
 
-      if (read_state_.unsplittable_)
+      if (read_state_.unsplittable_) {
+        if (offsets_extra_element_) {
+          RETURN_NOT_OK(add_extra_offset());
+        }
         return Status::Ok();
+      }
     } else {
       bool no_results = this->no_results();
 
       // Need to reset unsplittable if the results fit after all
       if (!no_results)
         read_state_.unsplittable_ = false;
 
-      if (!no_results || read_state_.done())
+      if (!no_results || read_state_.done()) {
+        if (offsets_extra_element_) {
+          RETURN_NOT_OK(add_extra_offset());
+        }
         return Status::Ok();
+      }
```

Comment:
This logic looks 100% correct to me, but I want to cut down on the number of branches in this routine since it's a little difficult to follow as it currently is.

Let's add a new function, `Status Reader::complete_read_loop`, where we can place any sort of post-read logic like this. So I think we can replace these two `return Status::Ok();` statements with `return complete_read_loop();`.

```
Status Reader::complete_read_loop() {
  if (offsets_extra_element_) {
    RETURN_NOT_OK(add_extra_offset());
  }

  return Status::Ok();
}
```